core/src/avm1/activation.rs

```diff
@@ -585,12 +585,20 @@ impl<'a, 'gc, 'gc_context> Activation<'a, 'gc, 'gc_context> {
         // ECMA-262 s. 11.6.1
         let a = self.context.avm1.pop();
         let b = self.context.avm1.pop();
-        let result: Value<'_> = if let Value::String(a) = a {
-            AvmString::concat(self.context.gc_context, b.coerce_to_string(self)?, a).into()
-        } else if let Value::String(b) = b {
-            AvmString::concat(self.context.gc_context, b, a.coerce_to_string(self)?).into()
-        } else {
-            (b.coerce_to_f64(self)? + a.coerce_to_f64(self)?).into()
+        let result: Value<'_> = match (a, b) {
+            (Value::String(a), b) => AvmString::concat(
+                self.context.gc_context,
+                b.to_primitive(self)?.coerce_to_string(self)?,
+                a,
+            )
+            .into(),
+            (a, Value::String(b)) => AvmString::concat(
+                self.context.gc_context,
+                b,
+                a.to_primitive(self)?.coerce_to_string(self)?,
+            )
+            .into(),
```

Comment:
The `coerce_to_string` feels a little redundant here, perhaps can lift it into `to_primitive` and rename `to_primitive` to something else.